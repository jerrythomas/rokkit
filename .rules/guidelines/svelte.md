# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
