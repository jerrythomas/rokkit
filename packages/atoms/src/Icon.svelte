<script>
	import { createEmitter } from '@rokkit/core'
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} name
	 * @property {any} [state]
	 * @property {string} [size]
	 * @property {string} [role]
	 * @property {any} [label]
	 * @property {boolean} [disabled]
	 * @property {number} [tabindex]
	 * @property {any} [checked]
	 * @event {MouseEvent} [onclick]
	 * @event {CustomEvent} [onchange]
	 * @event {MouseEvent} [onmouseenter]
	 * @event {MouseEvent} [onmouseleave]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		name,
		state = null,
		size = 'base',
		role = 'img',
		label = name,
		disabled = false,
		tabindex = $bindable(0),
		checked = $bindable(null),
		...events
	} = $props()

	let emitter = $derived(createEmitter(events, ['click', 'change', 'mouseenter', 'mouseleave']))
	function handleClick(e) {
		e.preventDefault()
		e.stopPropagation()

		if (!disabled) {
			if (role === 'checkbox' || role === 'option') {
				checked = !checked
				emitter.change(checked)
			}
			emitter.click()
		}
	}

	let validatedTabindex = $derived(role === 'img' || disabled ? -1 : tabindex)
	let ariaChecked = $derived(
		['checkbox', 'option'].includes(role) ? (checked !== null ? checked : false) : null
	)

	let small = $derived(size === 'small' || className.includes('small'))
	let medium = $derived(size === 'medium' || className.includes('medium'))
	let large = $derived(size === 'large' || className.includes('large'))
</script>

<rkt-icon
	class="flex flex-shrink-0 items-center justify-center {className}"
	class:small
	class:medium
	class:large
	class:disabled
	{role}
	aria-label={label}
	aria-checked={ariaChecked}
	onclick={handleClick}
	onkeydown={(e) => e.key === 'Enter' && e.currentTarget.click()}
	data-state={state}
	tabindex={validatedTabindex}
	onmouseenter={emitter.mouseenter}
	onnmouseleave={emitter.nmouseleave}
>
	<i class={name} aria-hidden="true"></i>
</rkt-icon>
