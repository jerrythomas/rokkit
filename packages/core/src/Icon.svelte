<script>
	export let name = ''
	let h
</script>

<square-icon
	class="flex flex-col flex-shrink-0 h-full items-center justify-center"
	bind:clientHeight={h}
	style:width="{h}px"
>
	<icon class={name} aria-hidden="true" />
</square-icon>
