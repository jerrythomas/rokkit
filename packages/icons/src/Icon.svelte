<script>
	import { toPascalCase } from '@sparsh-ui/core'
	import { iconStore } from './store.js'

	export let size = '1.5em'
	export let name
	export let title = name
	export let fill = 'none'
	export let stroke = 'currentColor'
	export let isActive = false
	export let icon = $iconStore[toPascalCase(name)]
</script>

<svg
	xmlns="http://www.w3.org/2000/svg"
	class="{$$props.class ? $$props.class : ''} icon"
	class:isActive
	{fill}
	{stroke}
	height={size}
	width={size}
	viewBox="0 0 24 24"
	on:hover
	on:mouseover
	on:focus
	on:click
>
	<title>{title}</title>
	<svelte:component this={icon} />
</svg>

<style>
	svg {
		display: flex;
		margin: auto;
		flex-shrink: 0;
		flex-grow: 0;
	}
	svg:focus {
		outline: none;
	}
</style>
