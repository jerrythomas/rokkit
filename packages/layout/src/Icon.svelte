<script>
	import { omit } from 'ramda'
	import { iconStore } from '../stores'

	export let size = '1.5em'
	export let name = ''
	export let fill = 'none'
	export let stroke = 'currentColor'

	export let icon = $iconStore[name]
	$: props = omit(['class'], $$restProps)
</script>

{#if typeof icon === 'string'}
	<icon
		class="{icon} {$$restProps.class || ''}"
		{...props}
		on:hover
		on:mouseover
		on:focus
		on:click
	/>
{:else}
	<svg
		xmlns="http://www.w3.org/2000/svg"
		class="{$$restProps.class || ''} icon"
		height={size}
		width={size}
		viewBox="0 0 24 24"
		{fill}
		{stroke}
		{...props}
		on:hover
		on:mouseover
		on:focus
		on:click
	>
		<svelte:component this={icon} />
	</svg>
{/if}

<style>
	svg {
		display: flex;
		margin: auto;
		flex-shrink: 0;
		flex-grow: 0;
	}
	svg:focus {
		outline: none;
	}
</style>
