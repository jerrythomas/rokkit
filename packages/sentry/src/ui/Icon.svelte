<script>
	export let size = '1.5em'
	export let title = ''
	export let icon
</script>

<svg
	xmlns="http://www.w3.org/2000/svg"
	class={$$props.class || ''}
	fill="none"
	height={size}
	width={size}
	viewBox="0 0 24 24"
>
	<title>{title}</title>
	<svelte:component this={icon} />
</svg>

<style>
	svg {
		margin: auto;
		flex-grow: 0;
		flex-shrink: 0;
		stroke: currentColor;
	}
</style>
