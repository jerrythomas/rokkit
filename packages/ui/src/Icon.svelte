<script>
	import { createEmitter } from '@rokkit/core'
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} name
	 * @property {any} [state]
	 * @property {string} [size]
	 * @property {string} [role]
	 * @property {any} [label]
	 * @property {boolean} [disabled]
	 * @property {number} [tabindex]
	 * @property {any} [checked]
	 * @event {MouseEvent} [onclick]
	 * @event {CustomEvent} [onchange]
	 * @event {MouseEvent} [onmouseenter]
	 * @event {MouseEvent} [onmouseleave]
	 */

	/** @type {Props} */
	let {
		class: classes = '',
		name,
		state = null,
		size = 'base',
		role = 'img',
		label = null,
		disabled = false,
		tabindex = $bindable(0),
		checked = $bindable(null),
		...events
	} = $props()

	let emitter = $derived(createEmitter(events, ['click', 'change', 'mouseenter', 'mouseleave']))
	function handleClick(e) {
		e.preventDefault()

		if (!disabled) {
			if (role === 'checkbox' || role === 'option') {
				checked = !checked
				emitter?.change(checked)
			}
			emitter?.click()
		}
	}

	let validatedTabindex = $derived(role === 'img' || disabled ? -1 : tabindex)
	let ariaChecked = $derived(
		['checkbox', 'option'].includes(role) ? (checked !== null ? checked : false) : null
	)
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<rk-icon
	class={classes}
	class:small={size === 'small'}
	class:medium={size === 'medium'}
	class:large={size === 'large'}
	class:disabled
	{role}
	aria-label={label ?? name}
	aria-checked={ariaChecked}
	onclick={handleClick}
	onkeydown={(e) => e.key === 'Enter' && e.currentTarget.click()}
	data-state={state}
	data-tag="icon"
	tabindex={validatedTabindex}
	onmouseenter={emitter.mouseenter}
	onnmouseleave={emitter.nmouseleave}
>
	<i class={name} aria-hidden="true"></i>
</rk-icon>
