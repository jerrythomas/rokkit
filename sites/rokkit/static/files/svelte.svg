<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
