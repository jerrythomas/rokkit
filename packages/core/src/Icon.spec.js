import { describe, it, expect, vi } from 'vitest'
import { render, fireEvent } from '@testing-library/svelte'
import Icon from './Icon.svelte'

describe('Icon component', () => {
	it('renders the icon with the provided name', () => {
		const { container } = render(Icon, { name: 'i-rokkit:mode-dark' })
		expect(container).toMatchSnapshot()
	})

	it('renders the icon with the provided name and label', () => {
		const { container } = render(Icon, {
			name: 'i-rokkit:mode-dark',
			label: 'Dark mode'
		})
		expect(container).toMatchSnapshot()
	})

	it('renders the icon with a custom class', () => {
		const { container } = render(Icon, {
			name: 'i-rokkit:mode-dark',
			class: 'custom-class'
		})
		expect(container).toMatchSnapshot()
	})

	it('renders the icon as disabled', () => {
		const { container } = render(Icon, {
			name: 'i-rokkit:mode-dark',
			disabled: true
		})
		expect(container).toMatchSnapshot()
	})

	it.each(['small', 'medium', 'large'])(
		'renders the icon with the specified size',
		(size) => {
			const { container } = render(Icon, {
				name: 'i-rokkit:mode-dark',
				size
			})
			expect(container).toMatchSnapshot()
		}
	)

	it.each(['small', 'medium', 'large'])(
		'renders the icon with the size when specified in class',
		(size) => {
			const { container } = render(Icon, {
				name: 'i-rokkit:mode-dark',
				class: size
			})
			expect(container).toMatchSnapshot()
		}
	)

	it('renders the icon with the specified role', () => {
		const { container } = render(Icon, {
			name: 'i-rokkit:mode-dark',
			role: 'button'
		})
		expect(container).toMatchSnapshot()
	})

	it('calls the click event when the icon is clicked', async () => {
		const handleClick = vi.fn()
		const { component, container } = render(Icon, {
			name: 'i-rokkit:mode-dark',
			role: 'button'
		})
		expect(container).toMatchSnapshot()
		component.$on('click', handleClick)
		const iconWrapper = container.querySelector('[role="button"]')
		await fireEvent.click(iconWrapper)

		expect(handleClick).toHaveBeenCalled()
	})

	it('calls the click event when the icon is focused and Enter key is pressed', async () => {
		const handleClick = vi.fn()
		const { getByRole, component } = render(Icon, {
			name: 'i-rokkit:mode-dark',
			role: 'button'
		})

		component.$on('click', handleClick)
		const iconWrapper = getByRole('button')
		await fireEvent.keyDown(iconWrapper, { key: 'Enter' })

		expect(handleClick).toHaveBeenCalled()
	})
})
