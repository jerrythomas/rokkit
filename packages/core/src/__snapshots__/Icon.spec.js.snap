// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon component > calls the click event when the icon is clicked 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center"
      role="button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon as disabled 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center disabled"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with a custom class 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center custom-class"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the provided name 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the provided name and label 1`] = `
<body>
  <div>
    <icon
      aria-label="Dark mode"
      class="flex flex-shrink-0 items-center justify-center"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the size when specified in class 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center small"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the size when specified in class 2`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center medium"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the size when specified in class 3`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center large"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the specified role 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center"
      role="button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the specified size 1`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center small"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the specified size 2`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center medium"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;

exports[`Icon component > renders the icon with the specified size 3`] = `
<body>
  <div>
    <icon
      aria-label="i-rokkit:mode-dark"
      class="flex flex-shrink-0 items-center justify-center large"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="i-rokkit:mode-dark"
      />
    </icon>
  </div>
</body>
`;
